src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -857,7 +915,9 @@ private static SizeAndAlignment ComputeInstanceSize(MetadataType type, LayoutInt
             {
                 if (type.IsValueType)
                 {
-                    instanceSize = LayoutInt.AlignUp(instanceSize, alignment, target);
+                    instanceSize = LayoutInt.AlignUp(instanceSize,
+                        alignUpInstanceByteSize ? alignment : LayoutInt.Min(alignment, target.LayoutPointerSize),
```

Comment:
I assume that the `alignUpInstanceByteSize = false` case will kick in for a struct like this on ARM32: 
```
    [StructLayout(LayoutKind.Explicit)]
    internal struct S3
    {
        [FieldOffset(0)]
        public ulong tmp1;
        [FieldOffset(8)]
        public Object tmp2;
    }
```
The algorithm will compute the instanceSize of this struct as 12. Is that correct?

It is a bug in the type loader. The instance size of this struct on ARM should be 16, so that the long field is aligned at 8 bytes, so that potential atomic 64-bit operations work fine on it. It would be better to fix the type loader instead of replicating the bug here.